feat(docs): add consistent hash document (#3499)
singularity-data/risingwave: docs/consistent-hash.md

![optimal data redistribution](./images/consistent-hash/data-redistribution-1.svg)

To minimize data movement when scaling occurs, we should be careful when we modify the vnode mapping. Below is an opposite example: modifying vnode mapping like this will result in $\frac{1}{2}$ of the data being moved.

![worst data redistribution](./images/consistent-hash/data-redistribution-2.svg)

### Streaming

We know that a fragment has several actors as its different parallelisms, and that upstream actors will send data to downstream actors via [dispatcher](./streaming-overview.md#actors). The figure below illustrates how actors distribute data based on consistent hash by example. 

![actor data distribution](./images/consistent-hash/actor-data.svg)

In the figure, we can see that one upstream actor dispatches data to three downstream actors. The downstream actors are scheduled on the parallel units mentioned in previous example respectively. 

> Based on our consistent hash design, the dispatcher is informed of the latest vnode mapping by meta node. It then decides how to send data by following steps:

Reviewer: Is it possible that vnode mapping in CN is stale? How meta inform the latest to CN?

Author: This is, in fact, not implemented yet. When scaling occurs, actors are designed to be dropped and recreated (currently), so the dispatchers would know the latest vnode mapping accordingly.

> Based on our consistent hash design, the dispatcher is informed of the latest vnode mapping by meta. It then decides how to send data by following steps:

Reviewer 2: ```suggestion
Based on our consistent hash design, the dispatcher is informed of the latest vnode mapping by meta node. It then decides how to send data by following steps:
```

1. Compute vnode of the data via the hash function $H$. Let the vnode be $v_k$.
2. Look up vnode mapping and find out parallel unit $p_n$ that vnode $v_k$ maps to. 
3. Send data to the downstream actor that is scheduled on parallel unit $p_n$ (remember that one actor will be scheduled on exactly one parallel unit).

In this way, all actors' data (i.e. actors' states) will be distributed according to the vnode mapping constructed by meta. 

When scaling occurs, actors will be re-scheduled accordingly. By modifying the vnode mapping in meta and make streaming act on the new vnode mapping, we could minimize data movement from following aspects:

- The data of existing actors will not be displaced too much. 
- The block cache of a compute node will not be invalidated too much.

### Batch

When we perform parallel batch read, we should partition the data for each parallelism in a certain way. Now that we have vnodes corresponding to disjoint sets of data, this naturally forms a data partition pattern. That is to say, one vnode could be viewed as a minimal data partition unit, and we could aggregate several vnodes together to get a larger data partition.

> In vnode mapping, one parallel unit will correspond to several vnodes. Therefore, we could use parallel unit as the partition bound. Data falling to one parallel unit in vnode mapping will be one partition. This is better than range partition in that the partition is more stable when the primary key of a materialized view distributes non-randomly.

Reviewer: > Therefore, we could use parallel unit as the partition bound.

What is use parallel unit as partition bound?

Author: The vnodes that mapped to one parallel unit will form a partition group. Lemme give a clearer description about this in the doc.

### Storage

If we look into the read-write pattern of streaming actors, we'll find that in most cases, actors only need to read the data written by itself (i.e. actor's internal states). That is to say, read data with the same vnodes as it previously writes. 

> Therefore, an instinctive way to place data in storage is to **group data by vnodes**. In this way, when actors perform read operation, it could touch as few SSTs as possible and thus trigger less I/O. 

Reviewer 2: ```suggestion
Therefore, an instinctive way to place data in storage is to **group data by vnodes**. In this way, when actors perform read operation, they could touch as few SST blocks as possible and thus trigger less I/O. 
```

#### Data Distribution

Here comes the main part, where we will construct a mapping that determines data distribution.

For all data $k \in U_k$, where $U_k$ is an unbounded set, we apply a hash function $v = H(k)$, where $v$ falls to a limited range. The hash function $H$ ensures that all data are hashed **uniformly** to that range. We call $v$ vnode, namely virtual node, as is shown as the squares in the figure below.

![initial data distribution](./images/consistent-hash/data-distribution.svg)

Then we have vnode mapping, which ensures that vnodes are mapped evenly to parallel units in the cluster. In other words, the number of vnodes that are mapped to each parallel unit should be as close as possible. This is denoted by different colors in the figure above. As is depicted, we have 3 parallel units (shown as circles), each taking $\frac{1}{3}$ of total vnodes. Vnode mapping is [constructed and maintained by meta](https://github.com/singularity-data/risingwave/blob/main/src/meta/src/manager/hash_mapping.rs).

> As long as the hash function $H$ could ensure uniformity, the data distribution determined by this strategy would be even across physical resources, even if data in $U_k$ skew to a certain range. 

Reviewer 3: Minor rewording suggestion: "even if data in ... skew to a certain range" ----> "even if data in ... are skewed to a certain extent."

Author: 1. Why use passive voice here? 🤔
2. "Range" here is sth like a line segment on a number axis. By "to a certain extent" it's more like "程度", which is not what I mean here.

Reviewer 3: Good question. When used as a verb, "skew" generally means causes something to be asymmetrical. In this sentence, data can only be the ones that are "caused" to appear in the way as-is. That's my understanding. I may be wrong. :) 

I misunderstood #2. Please ignore my comment.

Author: > data can only be the ones that are "caused" to appear in the way as-is.

I'll fix this. 

And I'll add an example to make clearer the meaning of "range".

We know that [Hummock](./state-store-overview.md#overview), our LSM-Tree-based storage engine, sorts key-value pairs by the order of the key. Hence, in order to group data by vnode on the basis of Hummock, we **encode vnode into the storage key**. The storage key will look like
```
table_id | vnode | ...
```
where `table_id` denotes the [state table](./storing-state-using-relational-table.md#relational-table-layer), and `vnode` is computed via $H$ on key of the data. 

To illustrate this, let's revisit the [previous example](#streaming). Executors of an operator will share the same logical state table, just as is shown in the figure below:

![actor state table](./images/consistent-hash/actor-state-table.svg)

Now that we have 12 vnodes in total in the example, the data layout in storage will accordingly look like this:

![storage data layout](./images/consistent-hash/storage-data-layout.svg)

Note that we only show the logical sequence and aggregation of data in this illustration. The actual data may be separated into different SSTs in Hummock.

> Since the way that certain data are hashed to vnode is invariant, the encoding of the data will also be invariant. How we schedule the fragment (e.g. parallelism of the fragment) will not affect data encoding. In other words, storage will not care about vnode mapping, which is determined by meta and used only by streaming. This is actually a way of decoupling compute layer and storage layer.

Reviewer 3: Another minor rewording suggestion: the last sentence-----> "This is actually a way of decoupling the compute layer from the storage layer."
